issubset: don't create a Set out of dicts

rfourquet opened the pull request:
Dictionaries also are hash tables, with fast `in`.
I also allowed myself a bit of code-densification.

rfourquet committed:
issubset: don't create a Set out of dicts

rfourquet commented:
This is easily backportable back to Julia 1.0.

stevengj commented:
Ranges also have fast `in`.  Maybe there should be a trait for this?

rfourquet commented:
> Maybe there should be a trait for this?

I wondered, I was thinking also of sorted vectors in general, but we can certainly write a trait including only ranges for now.

rfourquet commented:
This is now updated with a trait `hasfastin`.

rfourquet commented:
@nanosoldier runbenchmarks(ALL, vs = ":master")

rfourquet commented:
@nanosoldier `runbenchmarks(ALL, vs = ":master")`

KristofferC commented on base/abstractset.jl:
```diff
@@ -229,14 +229,22 @@ end
-        #sampling using these two methods (see #26198)
-        lenthresh = 70
+"""
+    hasfastin(T)
```
Can we just make this `_hasfastin` (or make the docstring into a comment) in order to clarify that this is currently just an internal way of structuring the code.

rfourquet replied:
I would say this is more than just an internal way of structuring the code, it's an extensible trait. So I'm quite relunctant to write it `_hasfastin`. Internal methods don't have to start with an underscore. I also don't understand removing the docstring, as even with it this function is unexported and not documented in the manual. There is other code in base which is not public but with docstring (and without underscores).

KristofferC replied:
>  as even with it this function is unexported and not documented in the manual

Fair enough.

staticfloat replied:
IMO there's never a reason to not have a docstring for a function, public or not.  It's very helpful for exploring internals.

rfourquet commented:
@nanosoldier `runbenchmarks(ALL, vs = ":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/32003#issuecomment-496470214) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/aebfb67504f3834d34e65cf11e9064d01627de95/2b711ac_vs_6ccd5c4/report.md). cc @ararslan

rfourquet commented:
I realise now that we don't have benchmarks in `BaseBenchmarks.jl` for `issubset` on dictionaries, and I could find `isubset` used in base only once (in `hasmethod`, which works on arrays I think), so runnning nanosoldier was probably irrelevant.

This would be good to go on my side, someone willing to review?

rfourquet commented:
Bump

nalimilan commented on base/abstractset.jl:
```diff
@@ -229,14 +229,22 @@ end
-        if rlen > lenthresh && !isa(r, AbstractSet)
+Determine whether the computation `x ∈ collection` where `collection::T` can be considered
+as a "fast" operation (typically constant or logarithmic complexity).
+The definition hasfastin(x) = hasfastin(typeof(x)) is provided for convenience so that instances
```
```suggestion
The definition `hasfastin(x) = hasfastin(typeof(x))` is provided for convenience so that instances
```

nalimilan reviewed (approved):
Looks good to me, assuming that existing tests cover all types included in the trait.

rfourquet committed:
add some tests for issubset on Dict

rfourquet commented:
@nalimilan Thanks for the review! I rebased and realized that the code of `issubset` had slightly changed, would you mind having again a quick look to double check that my rebase is correct? 

> assuming that existing tests cover all types included in the trait.

Good point, I added few basic tests for `issubset` on `Dict`, which was not tested at all.

nalimilan commented on base/abstractset.jl:
```diff
@@ -251,17 +251,16 @@ true
+        if !hasfastin(r)
+            # if r is big enough, convert it to a Set
+            # this threshold was empirically determined (cf. #26198)
+            rlen > 70 && return issubset(l, Set(r))
```
Why not put this in the `if` line above?

rfourquet replied:
I would say for the same reason as the previous check with `if isa(l, AbstractSet)`, to allow commenting between the comments. But I kinda agree that two nested simple `if` should be combined, I will update.

nalimilan commented on test/sets.jl:
```diff
@@ -347,6 +347,24 @@ end
+
+    # Base.hasfastin
+    @test all(Base.hasfastin, Any[Dict(1=>2), Set(1), BitSet(1), 1:9, 1:2:9,
+                               Dict, Set, BitSet, UnitRange, StepRange])
```
```suggestion
                                  Dict, Set, BitSet, UnitRange, StepRange])
```

nalimilan commented on test/sets.jl:
```diff
@@ -347,6 +347,24 @@ end
+    d2 = Dict(1=>nothing, 3=>nothing)
+    d3 = Dict(1=>nothing, 2=>nothing, 3=>nothing)
+    @test d3 == merge(d1, d2)
+    @test !(issubset(d1, d2))
```
```suggestion
    @test !issubset(d1, d2)
```
and so on.
